RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.cu

```diff
@@ -451,56 +450,86 @@ namespace pixelgpudetails {
     }  // end of loop (gIndex < end)
 
   }  // end of Raw to Digi kernel
-
+  template <bool isPhase2>
   __global__ void fillHitsModuleStart(uint32_t const *__restrict__ clusInModule,
                                       uint32_t *__restrict__ moduleStart,
                                       uint32_t const *__restrict__ nModules,
                                       uint32_t *__restrict__ nModules_Clusters) {
-    assert(gpuClustering::maxNumModules < 2048);  // easy to extend at least till 32*1024
+    constexpr int nMaxModules = isPhase2 ? phase2PixelTopology::numberOfModules : phase1PixelTopology::numberOfModules;
+    constexpr int startBPIX2 = isPhase2 ? phase2PixelTopology::layerStart[1] : phase1PixelTopology::layerStart[1];
+
+    assert(nMaxModules < phase2PixelTopology::numberOfModules);
+    assert(startBPIX2 < nMaxModules);
+    assert(nMaxModules < 4096);  // easy to extend at least till 32*1024
+    assert(nMaxModules > 1024);
+
     assert(1 == gridDim.x);
     assert(0 == blockIdx.x);
 
     int first = threadIdx.x;
 
-    // limit to maxHitsInModule()
-    for (int i = first, iend = gpuClustering::maxNumModules; i < iend; i += blockDim.x) {
+    // limit to MaxHitsInModule;
+    for (int i = first, iend = nMaxModules; i < iend; i += blockDim.x) {
       moduleStart[i + 1] = std::min(gpuClustering::maxHitsInModule(), clusInModule[i]);
     }
 
-    __shared__ uint32_t ws[32];
+    __shared__ uint32_t ws[64];
     cms::cuda::blockPrefixScan(moduleStart + 1, moduleStart + 1, 1024, ws);
-    cms::cuda::blockPrefixScan(moduleStart + 1025, moduleStart + 1025, gpuClustering::maxNumModules - 1024, ws);
+    constexpr int lastModules = isPhase2 ? 1024 : nMaxModules - 1024;
+    cms::cuda::blockPrefixScan(moduleStart + 1024 + 1, moduleStart + 1024 + 1, lastModules, ws);
 
-    for (int i = first + 1025, iend = gpuClustering::maxNumModules + 1; i < iend; i += blockDim.x) {
+    if constexpr (isPhase2) {
+      cms::cuda::blockPrefixScan(moduleStart + 2048 + 1, moduleStart + 2048 + 1, 1024, ws);
+      cms::cuda::blockPrefixScan(moduleStart + 3072 + 1, moduleStart + 3072 + 1, nMaxModules - 3072, ws);
+    }
+
+    for (int i = first + 1025, iend = isPhase2 ? 2049 : nMaxModules + 1; i < iend; i += blockDim.x) {
       moduleStart[i] += moduleStart[1024];
     }
     __syncthreads();
 
+    if constexpr (isPhase2) {
+      for (int i = first + 2049, iend = 3073; i < iend; i += blockDim.x) {
+        moduleStart[i] += moduleStart[2048];
+      }
+      __syncthreads();
+      for (int i = first + 3073, iend = nMaxModules + 1; i < iend; i += blockDim.x) {
+        moduleStart[i] += moduleStart[3072];
+      }
+      __syncthreads();
+    }
```

Comment:
I would find it easier to follow if the Phase-1 and Phase-2 parts were more clearly separated:
```c++
if constexpr (not isPhase2) {
    cms::cuda::blockPrefixScan(moduleStart + 1, moduleStart + 1, 1024, ws);
    cms::cuda::blockPrefixScan(moduleStart + 1024 + 1, moduleStart + 1024 + 1, nMaxModules - 1024, ws);
    for (int i = first + 1024 + 1, iend = nMaxModules + 1; i < iend; i += blockDim.x) {
      moduleStart[i] += moduleStart[1024];
    }    
    __syncthreads();
} else {
    cms::cuda::blockPrefixScan(moduleStart + 1, moduleStart + 1, 1024, ws);
    cms::cuda::blockPrefixScan(moduleStart + 1024 + 1, moduleStart + 1024 + 1, 1024, ws);
    cms::cuda::blockPrefixScan(moduleStart + 2048 + 1, moduleStart + 2048 + 1, 1024, ws);
    cms::cuda::blockPrefixScan(moduleStart + 3072 + 1, moduleStart + 3072 + 1, nMaxModules - 3072, ws);
    for (int i = first + 1024 + 1, iend = 2048 + 1; i < iend; i += blockDim.x) {
      moduleStart[i] += moduleStart[1024];
    }
    __syncthreads();
    for (int i = first + 2048 + 1, iend = 3072 + 1; i < iend; i += blockDim.x) {
      moduleStart[i] += moduleStart[2048];
    }
    __syncthreads();
    for (int i = first + 3072 + 1, iend = nMaxModules + 1; i += blockDim.x) {
      moduleStart[i] += moduleStart[3072];
    }
    __syncthreads();
}
```
(hopefully I did not introduce any bugs in the editing)